pkg/server/server.go

```diff
@@ -41,8 +49,38 @@ func Start(cfg Config) {
 		WriteTimeout: 40 * time.Second,
 	}
 
-	slog.Infof("listening on http://%s:%d", loopback, cfg.Port)
-	slog.Infof("targeting node %s", cfg.NodeIP)
+	network := "tcp"
+	addr := net.JoinHostPort(loopback, strconv.Itoa(cfg.Port))
+	if cfg.PreferUnixSocket {
+		network = "unix"
+		addr = cfg.UnixSocket
+	}
+
+	// gracefully shutdown HTTP server
+	idleConnClosed := make(chan struct{})
+	go func() {
+		sigCh := make(chan os.Signal, 1)
+		signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM, syscall.SIGQUIT)
+		slog.Info("Received signal: ", <-sigCh)
+		if err := httpServer.Shutdown(context.Background()); err != nil {
```

Comment:
Since the debug api calls are long running, I think adding a timeout before shutdown we could avoid closing in flight requests. Something like a 10-15sec timeout for handling in flight requests.

Reply:
Probably we can also have better signal on the in flight request by adding a few logs.

Reply:
Not sure I understood. `http.Server.Shutdown()` will 1) stop receiving the new connections, 2) wait indefinitely until the active connections become idle.  This is the graceful stop, we won't interrupt the active connections until they are done. They can take ~30 seconds, yes, but we want to finish them gracefully.
Another question is in which context this binary will be used. Kubernetes provides `terminationGracePeriodSeconds` field for the containers which is set to 30 seconds by default. This may be not enough for our binary as it may be in the middle of profiling request. `terminationGracePeriodSeconds` can indeed be increased to something like 45 seconds.

Reply:
Correct, what I was referring to is waiting indefinitely coupled with relying on `terminationGracePeriodSeconds` as a hard kill will work but it has reduced visibility on logs. Assuming we have `terminationGracePeriodSeconds`, we could set timeout on the context to be like 40sec along with a log.

Reply:
You mean, you'd like to see for how long we are waiting in the logs or code?

I came off the position that if:
- the binary is run by a human: the human has the control over the binary and can kill the process (at worst which I doubt to happen as profilings are long but not indefinite).
- the binary is run by Kubernetes: we let Kubernetes manage the graceful stop without interfering into this with our timeouts.

Since the second case is our target usage, where we run this binary the most: I didn't set any timeout.